### PR TITLE
Xcom gcs ds k8sexecutor

### PR DESCRIPTION
**Pull request by fhoda**
Tried to keep this one simple as well. Some things I left out were custom images for the tasks as this may be better presented in a KPO example.

- 2 Custom pod configs for 3 tasks (1 for load_data, and 1 for train and fit)
- Update Readme to include info on new example

### CODE REVIEW

File: dags/xcom_gcs_ds_k8sExecutor.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
+from typing import Container
```
Comment (josh-fell) [inline dags/xcom_gcs_ds_k8sExecutor.py]: This import looks to be unused and should be removed if true.

File: README.md [hunk 2]
```diff
@@ -9,4 +9,8 @@ In an effort to keep the examples easy to follow, much of the data processing an
 1. `xcom_gcs_ds.py` - A simple DS pipeline from data extraction to modeling.
     - Pulls data from BigQuery using the Google Provider (BigQueryHook) into a dataframe that preps, trains, and builds the model
     - Data is passed between the tasks using [XComs](https://airflow.apache.org/docs/apache-airflow/stable/concepts/xcoms.html)
-    - Uses GCS as an Xcom backend to easily track intermediary data in a scalable, external system
\ No newline at end of file
+    - Uses GCS as an Xcom backend to easily track intermediary data in a scalable, external system
+
+2. `xcom_gcs_ds_k8sExecutor.py` - A simple DS pipeline from data extraction to modeling that leverages the flexibility of the [Kubernetes Executor](https://www.astronomer.io/blog/new-kubernetesexecutor).
```
Comment (josh-fell) [inline README.md:14]: Might be worth mentioning that this pipeline should really _only_ be used with the `KubernetesExecutor` so others know the particular use case for this example vs. the original one. WDYT?

Comment (fhoda) [reply README.md:14]: For sure, will update.

File: dags/xcom_gcs_ds_k8sExecutor.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+from sklearn.model_selection import RepeatedStratifiedKFold
+from sklearn.model_selection import cross_val_score
+from lightgbm import LGBMClassifier
+
+
+docs = """
+By default, Airflow stores all return values in XCom. However, this can introduce complexity, as users then have to consider the size of data they are returning. Futhermore, since XComs are stored in the Airflow database by default, intermediary data is not easily accessible by external systems.
+By using an external XCom backend, users can easily push and pull all intermediary data generated in their DAG in GCS.
+
+Additionally using this with the Kubernetes Executor allows more fine grained controlled over the resources, images, and dependencies for each step on a per task basis.
```
Comment (santonaastronomer) [inline dags/xcom_gcs_ds_k8sExecutor.py]:
```suggestion
Additionally using this with the Kubernetes Executor allows more fine grained control over the resources, images, and dependencies for each step on a per task basis.
```

File: dags/xcom_gcs_ds_k8sExecutor.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+                                }
+                            )
+                            )
+                        ]
+                    )
+                )
+            }
+
+
+# Pod Config for train and fit tasks. This sets resource requests and limits for cpu and memory so that the Pod perform the tasks more quickly from minutes to seconds.
```
Comment (santonaastronomer) [inline dags/xcom_gcs_ds_k8sExecutor.py]:
```suggestion
# Pod Config for train and fit tasks. This sets resource requests and limits for cpu and memory so that the Pod can perform the tasks more quickly. Reduces execution time from minutes to seconds compared to without the use of `KubernetesExecutor`.
```

File: dags/xcom_gcs_ds_k8sExecutor.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+
+        # Drop redundant colulmn
+        df.drop(columns=['income_bracket_<=50K', 'marital_status', 'age'], inplace=True)
+
+        return df
+
+
+    @task(executor_config=modeling_config)
+    def train(df: pd.DataFrame):
```
Comment (santonaastronomer) [inline dags/xcom_gcs_ds_k8sExecutor.py]: Would it make more sense to call this function something like `validate` since it's not so much returning a trained model as a metric that is used to decide whether or not to use the model? I could see an extension of this step where versions of the model are checkpointed and the best iteration is saved/passed instead.

File: dags/xcom_gcs_ds_k8sExecutor.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+        """Train and validate model
+        
+        Returns accuracy score via XCom to GCS bucket.
+
+        Keyword arguments:
+        df -- data from previous step pulled from BigQuery to be processed. 
+        """
+
+        y = df['never_married'].values
+        X = df.drop(columns=['never_married']).values
```
Comment (santonaastronomer) [inline dags/xcom_gcs_ds_k8sExecutor.py]: I'd pull this step into `feature_engineering` as well, so that what that task returns is X and y -- then `train` and `fit` could be totally agnostic to actual attributes of the data, namely the 'never_married' column name.

**Review by santonaastronomer — APPROVED** [review]

File: dags/xcom_gcs_ds_k8sExecutor.py (hunk 5 above)
Comment (fhoda) [reply dags/xcom_gcs_ds_k8sExecutor.py]: Yes that would be a great add on. Do we want to make this change on the first example dag as well? We could even make it more descriptive as well like "cross_validation" if that make sense.

Comment (santonaastronomer) [reply dags/xcom_gcs_ds_k8sExecutor.py]: Yup, that makes sense.

File: dags/xcom_gcs_ds_k8sExecutor.py (hunk 6 above)
Comment (fhoda) [reply dags/xcom_gcs_ds_k8sExecutor.py]: @santonaastronomer To output two XComs from feature_engineering and then read them  back in I would have to output them as a JSON. There may be a more simpler way that I am overlooking, but something like this...

XCom Output from feature_engineering
```
return {'X': df.drop(columns=['never_married']).to_json(orient='index'), 'y': df['never_married'].to_json(orient='index')}

```

Then read then back in with `io.StringIO` on the downstream tasks. Which would look like this

```
y = pd.read_json(StringIO(df['y']), orient='index').values
X = pd.read_json(StringIO(df['X']), orient='index').values
```

This is not necessarily bad, but wanted to make sure here before committing anything.

Comment (santonaastronomer) [reply dags/xcom_gcs_ds_k8sExecutor.py]: @fhoda I think this is perfectly fine. In a deployed model scenario, the standard would be for one service to featurize data and write them to a featurestore. A /predict endpoint (ideally called by a different service) should be able to predict on anything from the featurestore – irrespective of whether or not there's targets to score the predictions against. Doing additional stuff to the input data in your train and fit functions feels against that vein.

Comment (fhoda) [reply dags/xcom_gcs_ds_k8sExecutor.py]: Done!

**Comment by fhoda** [comment]
@josh-fell please take a final look and approve accordingly. If you like you can go ahead and merge so that we can get this added quickly.

**Review by josh-fell — APPROVED** [review]
👍
